runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1086 @@
+			T::AssignSlotOrigin::ensure_origin(origin)?;
+
+			let manager = T::Registrar::manager_of(id).ok_or(Error::<T>::ParaDoesntExist)?;
+
+			ensure!(T::Registrar::is_parathread(id), Error::<T>::NotParathread,);
+
+			ensure!(
+				!Self::has_permanent_slot(id) && !Self::has_temporary_slot(id),
+				Error::<T>::SlotAlreadyAssigned
+			);
+
+			let current_lease_period: T::BlockNumber = Self::lease_period_index();
+			ensure!(
+				!T::Leaser::already_leased(
+					id,
+					current_lease_period,
+					// Check current lease & next one
+					current_lease_period.saturating_add(
+						T::BlockNumber::from(2u32)
+							.saturating_mul(T::PermanentSlotLeasePeriodLength::get().into())
+					)
+				),
+				Error::<T>::OngoingLeaseExists
+			);
+
+			ensure!(
+				PermanentSlotCount::<T>::get() + 1 <= T::MaxPermanentSlots::get(),
+				Error::<T>::MaxPermanentSlotsExceeded
+			);
+
+			<PermanentSlotCount<T>>::try_mutate(|count| -> DispatchResult {
+				Self::configure_slot_lease(
+					id,
+					manager,
+					current_lease_period,
+					T::PermanentSlotLeasePeriodLength::get().into(),
+				)
+				.map_err(|_| Error::<T>::CannotUpgrade)?;
+				PermanentSlots::<T>::insert(id, Some(current_lease_period));
+				*count = count.checked_add(One::one()).ok_or(ArithmeticError::Overflow)?;
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::PermanentSlotAssigned(id));
+			Ok(())
+		}
+
+		/// Assign a temporary parachain slot
+		#[pallet::weight((1_000, DispatchClass::Operational))]
+		pub fn assign_temp_parachain_slot(
+			origin: OriginFor<T>,
+			id: ParaId,
+			lease_period_start: SlotLeasePeriodStart,
+		) -> DispatchResult {
+			T::AssignSlotOrigin::ensure_origin(origin)?;
+
+			let manager = T::Registrar::manager_of(id).ok_or(Error::<T>::ParaDoesntExist)?;
+
+			ensure!(T::Registrar::is_parathread(id), Error::<T>::NotParathread,);
+
+			ensure!(
+				!Self::has_permanent_slot(id) && !Self::has_temporary_slot(id),
+				Error::<T>::SlotAlreadyAssigned
+			);
+
+			let current_lease_period: T::BlockNumber = Self::lease_period_index();
+			ensure!(
+				!T::Leaser::already_leased(
+					id,
+					current_lease_period,
+					// Check current lease & next one
+					current_lease_period.saturating_add(
+						T::BlockNumber::from(2u32)
+							.saturating_mul(T::TemporarySlotLeasePeriodLength::get().into())
+					)
+				),
+				Error::<T>::OngoingLeaseExists
+			);
+
+			TemporarySlots::<T>::insert(
+				id,
+				Some(ParachainTemporarySlot {
+					manager,
+					period_begin: match lease_period_start {
+						SlotLeasePeriodStart::Current => current_lease_period,
+						SlotLeasePeriodStart::Next => current_lease_period + One::one(),
+					},
+					period_count: T::TemporarySlotLeasePeriodLength::get().into(),
+					last_lease: None,
+					lease_count: 0,
+				}),
+			);
+
+			if lease_period_start == SlotLeasePeriodStart::Current {
+				Self::allocate_temporary_slot_leases(current_lease_period)?;
```

Comment:
This can fail even though you've changed storage above. Are you sure this should not be atomic?